binary_search_tree/tree.py

```diff
@@ -3,57 +3,183 @@ def __init__(self, key, val = None):
+    # Space Complexity: O(1) because we are not allocating any memory
+    # self refers to object the class is making
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+      
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else: 
+                current = current.left
+        return None
+
+
+    # Time Complexity:  O(N).  Since it is recursive it is O(N) because the stack size will depend on the input size.  Behind the scenes a new stack frame is created to hold the return address and variables you create in the new call.
+    # Space Complexity: 0(N)  depends on input size
+
+    # first go to left 
+    # add to the list 
+    # go to the right
+    
+    
+
+        
+    def preorder_traverse(self, node, node_list): 
+        node_list.append(self.create_dict(node))
+        if node.left!=None:
+            self.preorder_traverse(node.left, node_list)
+        if node.right!=None:
+            self.preorder_traverse(node.right, node_list)
+
+    def postorder_traverse(self, node, node_list): 
+        if node.left!=None:
+            self.postorder_traverse(node.left, node_list)
+        if node.right!=None:
+            self.postorder_traverse(node.right, node_list)
+        node_list.append(self.create_dict(node))
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def inorder_traverse(self, node, node_list): 
+        if node.left!=None:
+            self.inorder_traverse(node.left, node_list)
+        node_list.append(self.create_dict(node))
+        if node.right!=None:
+            self.inorder_traverse(node.right, node_list)
+        
+       
+        
     def inorder(self):
-        pass
+        node_list = []
+        if self.root == None:
+            return []
+        self.inorder_traverse(self.root, node_list)  
+        return node_list
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: 0(N)
+    # Space Complexity: Space Complexity: 0(N)  depends on input size    
     def preorder(self):
-        pass
+        node_list = []
+
+        if self.root == None:
+            return []
+        self.preorder_traverse(self.root, node_list)  
+        return node_list
+
+     
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: 0(N)
+    # Space Complexity: 0(N) depends on input size        
     def postorder(self):
-        pass
+        node_list = []
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def height(self):
-        pass
+        if self.root == None:
+            return []
+        self.postorder_traverse(self.root, node_list)  
+        return node_list
+        
+
+    # Time Complexity: 0(1) constant time
+    # Space Complexity:  0(1)   
+    def height(self):   
```

Comment:
👍 this is O(n) for time complexity and O(n) space complexity if the tree is unbalanced and O(log n) if the tree is balanced.